Doc/library/__main__.rst

```diff
@@ -67,7 +67,32 @@ elsewhere::
     from echo import main as echo_main
 
     def echo_platform():
-       echo_main(sys.platform)
+        echo_main(sys.platform)
+
+The spirit of this design is inherited from the C programming language, where
+the function whose name is *main* is the entry-point of a program.  In C,
+*main* also returns an integer, which becomes the exit code of the process.
+Zero typically indicates successful termination, and other codes indicate some
+type of failure.  :func:`sys.exit` provides the API for exiting with an
+explicit exit code.  A popular convention in Python is for *main* functions to
```

Comment:
This convention is because that's what console scripts entry points do

Reply:
eg here https://github.com/pypa/pip/blob/7ec0fa5142466f8402bb866a1dcfb8ca6ffdc66b/src/pip/_vendor/distlib/scripts.py#L41

Reply:
> This convention...

Can you clarify what you mean by "this convention?" There are (at least) three
conventions at play in this paragraph:

1. Whether or not you should use `if __name__ == '__main__'` in a `__main__.py`
   file
2. Whether or not your main function should return an integer, like it would in
   C
3. Whether or not you should always use `sys.exit` to always return an explicit
   and intentional exit value

To my understanding, you might be suggesting several concrete changes, but
I'm not sure which:

- Do not discuss the historic role of _main_ in C because it is not the most
  relevant justification. The most relevant justification for this pattern are
  Python packaging reasons (console script entry points). **Discuss those
  reasons instead.**
  - If this is your argument, I'm confused, because you can tell console entry
    scripts to point to anything using this syntax: `<name> = [<package>.[<subpackage>.]]<module>[:<object>.<object>]`. See [Console Scripts](https://setuptools.readthedocs.io/en/latest/userguide/entry_point.html#console-scripts)
  - Thus, the console scripts use case does not _require_ a function named
    main, nor does it require an explicit exit code using `sys.exit` (the docs
    linked above do not use `sys.exit` in their example).
- All examples should include _main_ functions that return an integer and pass
  that integer to `sys.exit` explicitly.
  - If a python program exits from an exception, the exit code is `1`, so the
    convention of returning an explicit exit code is not strictly necessary,
    and more of a stylistic choice.
    - Therefore, _main_ returning an integer is a stylistic choice
    - Therefore, the existence of _main_ itself is a stylistic choice
    - Therefore, the "idiomatic usage" section should gently guide the user
      towards this best practice and explain the reasonning behind it, rather
      than strictly perscribing the user to follow it.
- Do not advise against using `if __name__ == '__main__'` in `__main__.py`.
  Although it may be redundant in some cases, it is still best practice to
  include it.
  - I have noticed that documentation typically tells users to include it
  - However, I do not understand why users should include it in most cases,
    because it seems to me that it is almost always redundant, except in the
    case of console scripts where `__name__` would be `package.__main__`. Even
    then, wouldn't you **not** want your code to be under a `__name__ == '__main__'` block in that case? If it were, the code you'd want to run
    wouldn't run when invoked as a console script. To me, this seems like
    another reason to tell users _not_ to guard code in `__main__.py` under an
    `if __name__ == '__main__'` block. `__main__.py` should always run from
    top-to-bottom; is that not the case?
  - That being said, this is also a matter of idiomatic usage. See
    1a9956cb9d2dbb2672ebcd4003375b1ae2cb4441, where I switched this bit to
    be a corresponding "idiomatic usage" section, and I also added a comment
    copying the notes here and contemplating deleting this whole paragraph.

Reply:
whenever I create a python module that I want to run from the console

I will *always* endeavour to end it with the following bytes:
```
b'\n\nif __name__ == "__main__":\n    sys.exit(main())\n'
```

Specifically I don't use `main(' '.join(sys.argv))` or `exit(main())` or `main()` or `sys.exit(trio.run(main))`


I do this so that I can use my script from either a entry point:

```ini
[options.entry_points]
console_scripts =
    path-to-my-package = path.to.my.package.__main__:main
```

or from runpy:
```sh
python -m path.to.my.package
```

and have broadly equivalent behaviour

This is because when I `pip` install such a package I get a file like this:

```python
#!/home/graingert/.virtualenvs/testing38/bin/python3.8
# -*- coding: utf-8 -*-
import re
import sys
from path.to.my.package.__main__ import main
if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw|\.exe)?$', '', sys.argv[0])
    sys.exit(main())
```

Reply:
also I think `' '.join(sys.argv)` is a security issue. Afaik you need `shlex.join` but I'm not sure. I think it's safest to leave it as a list

Reply:
Ok, I think I finally get what you're saying; see f095362efb2df6ddbe5cd6a8af62f2d5e10393ca

Thank you for your patience in helping me understand!!